include/libtorrent/aux_/rtc_stream.hpp

```diff
@@ -0,0 +1,343 @@
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_RTC_STREAM_HPP_INCLUDED
+#define TORRENT_RTC_STREAM_HPP_INCLUDED
+
+#include "libtorrent/aux_/packet_buffer.hpp"
+#include "libtorrent/close_reason.hpp"
+#include "libtorrent/error_code.hpp"
+#include "libtorrent/io.hpp"
+#include "libtorrent/io_context.hpp"
+#include "libtorrent/proxy_base.hpp"
+#include "libtorrent/time.hpp"
+#include "libtorrent/udp_socket.hpp"
+
+#include <functional>
+#include <memory>
+#include <queue>
+#include <vector>
+
+#ifndef BOOST_NO_EXCEPTIONS
+#include "libtorrent/aux_/disable_warnings_push.hpp"
+#include <boost/system/system_error.hpp>
+#include "libtorrent/aux_/disable_warnings_pop.hpp"
+#endif
+
+namespace rtc {
+	class PeerConnection;
+	class DataChannel;
+}
+
+namespace libtorrent {
+namespace aux {
+
+struct TORRENT_EXTRA_EXPORT rtc_stream_init
+{
+	std::shared_ptr<rtc::PeerConnection> peer_connection;
+	std::shared_ptr<rtc::DataChannel> data_channel;
+};
+
+// This is the user-level stream interface to WebRTC DataChannels.
+struct TORRENT_EXTRA_EXPORT rtc_stream
+{
+	using lowest_layer_type = rtc_stream;
+	using endpoint_type = tcp::socket::endpoint_type;
+	using protocol_type = tcp::socket::protocol_type;
+
+	using executor_type = tcp::socket::executor_type;
+	executor_type get_executor() { return m_io_context.get_executor(); }
+
+	explicit rtc_stream(io_context& ioc, rtc_stream_init const& init);
+	~rtc_stream();
+	rtc_stream& operator=(rtc_stream const&) = delete;
+	rtc_stream(rtc_stream const&) = delete;
+	rtc_stream& operator=(rtc_stream&&) noexcept = delete;
+	rtc_stream(rtc_stream&&) noexcept;
+
+	lowest_layer_type& lowest_layer() { return *this; }
+
+#ifndef BOOST_NO_EXCEPTIONS
+	template <class IO_Control_Command>
+	void io_control(IO_Control_Command&) {}
+#endif
+
+	template <class IO_Control_Command>
+	void io_control(IO_Control_Command&, error_code&) {}
+
+#ifndef BOOST_NO_EXCEPTIONS
+	void non_blocking(bool) {}
+#endif
+
+	void non_blocking(bool, error_code&) {}
+
+#ifndef BOOST_NO_EXCEPTIONS
+	void bind(endpoint_type const&) {}
+#endif
+
+	void bind(endpoint_type const&, error_code&) {}
+
+#ifndef BOOST_NO_EXCEPTIONS
+	template <class SettableSocketOption>
+	void set_option(SettableSocketOption const&) {}
+#endif
+
+	template <class SettableSocketOption>
+	void set_option(SettableSocketOption const&, error_code&) {}
+
+#ifndef BOOST_NO_EXCEPTIONS
+	template <class GettableSocketOption>
+	void get_option(GettableSocketOption&) {}
+#endif
+
+	template <class GettableSocketOption>
+	void get_option(GettableSocketOption&, error_code&) {}
+
+	void cancel(error_code&)
+	{
+		cancel_handlers(boost::asio::error::operation_aborted);
+	}
+
+	void close();
+	void close(error_code const&) { close(); }
+
+	close_reason_t get_close_reason();
+
+	bool is_open() const;
+
+	void add_read_buffer(void* buf, std::size_t len);
```

Comment:
I would suggest `add_read_buffer()` and `add_write_buffer()` take a `mutable_buffer` and `const_buffer` respectively. That way the call sites will be simpler (I think)

Reply:
I also think it would be reasonable to remove these functions as they are only used once and only contain two lines of code